src/python/pants/backend/python/tasks2/partition_targets.py

```diff
@@ -0,0 +1,146 @@
+    """Returns a subset containing the given targets.
+
+    Raises a KeyError if such target does not exist.
+    """
+    return self._subset_by_target[target]
+
+  def find_subset_for_targets(self, targets):
+    """Returns a single subset that contains all given targets.
+
+    If such subset does not exists within the partition a ValueError is raised.
+    """
+    if not self._subsets:
+      raise ValueError('No subsets.')
+    if not targets:
+      # Any subset will do.
+      return next(iter(self._subsets))
+    subset = self._subset_by_target.get(next(iter(targets)))
+    if not subset or not subset.issuperset(targets):
+      raise ValueError('No subset contains all targets: {}'.format(
+          ' '.join(tgt.address.spec for tgt in targets)))
+    return subset
+
+
+class PartitionTargets(Task):
+  """Generates partitions of the target roots.
+
+  This tasks produces `TargetPartition`s. Each representing a partition of the target roots into
+  non-empty disjoint subsets of targets based on a requested partitioning strategy.
+
+  There are several supported strategies to produce a TargetPartition from a set of target roots:
+
+  STRATEGY_MINIMAL:
+      partitions the target roots into the smallest possible number of subsets such that if one target
+      root depends on another, then both of them will be in the subset. This ensures
+      that incompatible target roots will fall into different subsets while minimizing the number
+      of chroots.
+  STRATEGY_PER_TARGET:
+      A partition in which Each target root will be in its own isolated subset.
+      This provides maximal isolation but can be slower if there are many target
+      roots.
+  STRATEGY_GLOBAL:
+      A partition with one single subset for all target roots. This means that all target
+      roots need to be compatible with each other.
+
+  The strategies that are requested via require_data() will be provided.  In addition a product
+  TARGETS_PARTITION is provided which is a map between the requested strategy and the
+  partition.
+  """
+
+  TARGETS_PARTITIONS = 'targets_partitions'
+  STRATEGY_MINIMAL = 'targets_partition_minimal'
+  STRATEGY_PER_TARGET = 'targets_partition_per_target'
+  STRATEGY_GLOBAL = 'targets_partition_global'
+
+  @classmethod
+  def product_types(cls):
+    return [cls.TARGETS_PARTITIONS,
+            cls.STRATEGY_MINIMAL,
+            cls.STRATEGY_PER_TARGET,
+            cls.STRATEGY_GLOBAL]
+
+  @classmethod
+  def register_options(cls, register):
+    super(PartitionTargets, cls).register_options(register)
+    return
+    register(
+        '--strategy',
+        choices=[cls.STRATEGY_MINIMAL, cls.STRATEGY_PER_TARGET, cls.STRATEGY_GLOBAL],
+        help='Specifies how to partition the targets into subsets.',
+        default=cls.STRATEGY_GLOBAL)
+
+  @classmethod
+  def _minimal_partition(cls, targets):
+    subsets_by_head = {}
+
+    closures = {target: target.closure() for target in targets}
```

Comment:
This is O(n^2) in the number of targets.  In the `./pants test ::` scenario this might be prohibitive. Off the top of my head it seems like an inverted topological sort would give you this in O(n)?